### PR TITLE
migrate to `tower-sessions`

### PR DESCRIPTION
**Pull request by maxcountryman**
We have entirely reimagined this crate's API, having now rewritten it to use `tower-sessions`.

The upshot of this is issues with deadlocks are a thing of the past. However, applications that rely on prior versions of the crate will have to evaluate the new API and decide if it's appropriate to migrate or not.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -2,141 +2,117 @@ name: Rust
...
+          #    cargo install cargo-tarpaulin
       - uses: Swatinem/rust-cache@v2
       - name: Run integration tests
-        uses: actions-rs/cargo@v1
-        with:
-          command: test
-          args: --manifest-path "axum-login-tests/Cargo.toml" --features ${{ matrix.database_backend }}
+        run: |
+          cargo test --test integration-test
+      #- name: Generate code coverage
```
Comment (czocher) [inline .github/workflows/rust.yml:78]: Maybe remove it?

File: README.md [hunk 2]
```diff
@@ -23,117 +23,56 @@ axum-login
...
+  for high-performing and ergonomic session management. _Look ma, no deadlocks!_
 
 ## 📦 Install
 
 To use the crate in your project, add the following to your `Cargo.toml` file:
 
 ```toml
 [dependencies]
-axum-login = "0.6.0"
+axum-login = "0.7.0"
```
Comment (czocher) [inline README.md:54]: Imho it's a major version change. Maybe 1.0.0-rc?

**Review by czocher — CHANGES_REQUESTED** [review]

File: examples/sqlite.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    ) -> Result<Option<Self::User>, Self::Error> {
+        let user: Option<Self::User> = sqlx::query_as("select * from users where username = ? ")
+            .bind(creds.username)
+            .fetch_optional(&self.pool)
+            .await?;
+
+        Ok(user.filter(|user| {
+            let parsed_hash =
+                PasswordHash::new(&user.password).expect("Failed to create password hash.");
+            Argon2::default()
```
[the remark this answers is not in the document]
Comment (maxcountryman) [reply examples/sqlite.rs:107]: Good call--thank you!

File: examples/sqlite.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
+use std::{fmt::Debug, net::SocketAddr};
```
[the remark this answers is not in the document]
Comment (maxcountryman) [reply examples/sqlite.rs:1]: This makes sense to me. I'm going to track this in a separate ticket and will plan to follow up after we merge this patch.

File: src/session.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            tracing::Span::current().record("user.id", user.id().to_string());
+        }
+
+        self.user = None;
+        self.data = Data::default();
+        self.session.flush();
+
+        self.update_session().map_err(Error::Session)?;
+
+        Ok(user.clone())
```
[the remark this answers is not in the document]
Comment (maxcountryman) [reply src/session.rs]: Great catch!

File: examples/sqlite.rs [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
```
[the remark this answers is not in the document]
Comment (maxcountryman) [reply examples/sqlite.rs:2]: Good call. I'll annotate with comments.

File: README.md (hunk 2 above)
Comment (maxcountryman) [reply README.md:54]: I'm comfortable moving to `1.0.0` with this API. However, I also anticipate we'll iterate a bit more on this before it's completely settled. Following [this semantic versioning spec](https://semver.org/#spec-item-4), I think we have more flexibility at `0.x.y` so my thinking is to wait a little longer.

File: src/session.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+    /// The authentication and authorization backend.
+    pub backend: Backend,
+
+    data: Data<UserId<Backend>>,
+    session: Session,
+}
+
+impl<Backend: AuthnBackend> AuthSession<Backend> {
+    const DATA_KEY: &'static str = "axum-login.data";
```
[the remark this answers is not in the document]
Comment (maxcountryman) [reply src/session.rs:79]: I'll add this as a follow up as well.

**Comment by maxcountryman** [comment]
Thanks for your fantastic feedback, @czocher. I'll go ahead and merge this now and start working on a release.

**Pull request merged**